### PR TITLE
Branches - Julia & Eve

### PR DESCRIPTION
**Pull request by tofuandeve**
# Adagrams
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the components that make up a method? | A method signature includes method name and parameters, a block of code that will be executed. A method can return a value (with return keyword) or nil (if method only prints output). In order for a method to execute, one must call/ invoke with arguments if parameters are set.
What are the advantages of using git when collaboratively working on one code base? | Git makes it easier for remotely updating/ maintaining code by multiple people.
What kind of relationship did you and your pair have with the unit tests? | We ran the tests before implementing code to see the tests fail. And as we wrote code, we ran tests again to make sure all tests pass. When we were done refactoring code, we ran tests again to make sure the tests still pass.
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? | We used .map, .select, .max_by and .min_by on Wave 4. Because they allowed us to easily pull values out of our data and convert data structures.
What was one method you and your pair used to debug code? | We used pry to debug variables with unexpected values on Wave 4. Reading stack trace to find where the bugs were originating. 
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? |  We figured out how to be more communicative and effective throughout the project as we figured out our workflow and communication styles.

### CODE REVIEW

**Comment by tildeee** [comment]
# Adagrams
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | x
Small commits with meaningful commit messages | x, I would prefer commit messages that are less "Implemented wave 2" and more "implements the score_word method" or "refactors the score table data structure" etc
**Code Requirements** |
`draw_letters` method | x
Uses appropriate data structure to store the letter distribution | x
All tests for `draw_letters` pass | x
`uses_available_letters?` method | x
All tests for `uses_available_letters?` pass | x
`score_word` method | x
Uses appropriate data structure to store the letter scores | x
All tests for `score_word` pass | x
`highest_score_from` method | x
Appropriately handles edge cases for tie-breaking logic | x
All tests for `highest_score_from` pass | x
**Overall** |



Eve and Julia!! You two have a fantastic solution! I see a lot of thought put into your submission; you all achieved not only a good, logical, readable solution, but it has obviously gone through refactoring to make it as clever and elegant as possible within this time.


Some of the only things I could think of for improvement on this submission would be to add in the unit tests for the optional wave 5, and practice using more descriptive commit messages.


Otherwise, please keep up the clarity in logic and code, and maintain the refactoring when you can. Great work on this submission!

File: lib/adagrams.rb [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+  letters = []
+  distribution.each do |letter, quantity|
+    quantity.times { letters << letter.to_s }
+  end
+  
+  # push 10 random letters to strings array, removing letter from letters array each time to prevent repitition
+  drawn_letters = []
+  10.times do
+    letters.shuffle!
+    drawn_letters << letters.pop
```
Comment (tildeee) [inline lib/adagrams.rb:45]: Nice! Using `.pop` here is concise and a good call

File: lib/adagrams.rb [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+# WAVE 2
+def uses_available_letters?(input, letters_in_hand)
+  current_letters_in_hand = letters_in_hand.dup
+  # split user input into array of chararacters
+  characters = input.upcase.split('')
+  # for each character, save index number to index variable
+  characters.each do |char|
+    index = current_letters_in_hand.find_index(char)
+    # if index number is truthy, delete element at index
+    if index
```
Comment (tildeee) [inline lib/adagrams.rb:60]: Nice logic here that relies on truthiness and understanding what `find_index` does!

File: lib/adagrams.rb [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+      current_letters_in_hand.delete_at(index)
+      # if index number is falsey, return false
+    else 
+      return false
```
Comment (tildeee) [inline lib/adagrams.rb:64]: Nice logic; putting a `return false` here will break the method out as soon as it needs to

File: lib/adagrams.rb [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+  
+  # save words with length of 10 to winner array
+  winner = high_scores.select { |element| element.length == 10 }
+  
+  # if there are no words with length of 10, assign first, shortest word to winner variable 
+  # if there are words with length of 10, assign first word to winner variable
+  winner = winner.empty? ? high_scores.min_by { |element| element.length } : winner[0]
+  
+  return { word: winner, score: best_result[:score] }
+end
```
Comment (tildeee) [inline lib/adagrams.rb:114]: I'm obsessed with y'alls implementation of this method! It's so clever and so concise! It looks so good! The only thought for improvement is maybe using a different variable name besides `obj`. Otherwise, I love how it reads!

File: lib/adagrams.rb [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+# WAVE 5
+
+def is_in_english_dict?(input)
```
Comment (tildeee) [inline lib/adagrams.rb:118]: Great work on the optional! :D ...Tests? ;)
